src/tensors/tensor.jl

```diff
@@ -157,89 +119,28 @@ Construct a `TensorMap` by explicitly specifying its block data.
 Alternatively, the domain and codomain can be specified by passing a [`HomSpace`](@ref)
 using the syntax `codomain ← domain` or `domain → codomain`.
 """
-function TensorMap(data::AbstractDict{<:Sector,<:DenseMatrix},
+function TensorMap(data::AbstractDict{<:Sector,<:AbstractMatrix},
                    V::TensorMapSpace{S,N₁,N₂}) where {S,N₁,N₂}
-    dom = domain(V)
-    codom = codomain(V)
-    I = sectortype(S)
-    I == keytype(data) || throw(SectorMismatch())
-    if I == Trivial
-        if dim(dom) != 0 && dim(codom) != 0
-            return TensorMap(data[Trivial()], codom, dom)
-        else
-            return TensorMap(valtype(data)(undef, dim(codom), dim(dom)), codom, dom)
-        end
-    end
-    blocksectoriterator = blocksectors(codom ← dom)
-    for c in blocksectoriterator
+    
+    T = eltype(valtype(data))
+    t = TensorMap{T}(undef, V)
+    for (c, b) in blocks(t)
         haskey(data, c) || throw(SectorMismatch("no data for block sector $c"))
+        datac = data[c]
+        size(datac) == size(b) || throw(DimensionMismatch("wrong size of block for sector $c"))
+        copy!(b, datac)
```

Comment:
If we instead have `setindex!(t::AbstractTensorMap, c::Sector)` we could make the size check and haskey check automatic

Reply:
I think I would indeed prefer this, mostly for the `setindex!` part, I like that we can avoid handing the user a view to then use in-place, and the ability to re-use some of the boundschecking functionality is also convenient. Additionally, I feel like it really is quite intuitive and a proper use of `getindex`/`setindex`, so I cannot really think of a reason not to do this, as we also do it for the fusiontrees